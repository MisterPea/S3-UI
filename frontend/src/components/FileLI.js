import * as React from 'react';
import { useState } from 'react';
import propTypes from 'prop-types';
import { IoEllipsisVerticalCircle, IoInformationCircleSharp } from 'react-icons/io5';
import { motion } from 'framer-motion';
import { Folder } from './graphic_elements/Icons';
import FileIcon from './helpers/FileIcon';
import FolderDropdown from './FolderDropdown';
import FolderDeleteModal from './FolderDeleteModal';
import ModalComponentWrapper from './ModalComponentWrapper';
import FileModal from './FileModal';
import fileSizeTruncate from './helpers/fileSizeTruncate';

/**
 * Component to render a folder or a file
 * @prop {string} type The type to be rendered `file` or `folder`
 * @prop {string} name Name associated with the file or folder
 * @prop {string} lastModified Date object was last modified (Only for files)
 * @prop {number} size Size of the file (Only for files)
 * @prop {string} filePath File path leading to children (Only for folders)
 * @prop {func?} folderClick Callback for folders that are clicked
 * @prop {string} locale Region of the bucker - passed to modal.
 * @prop {func?} uploadClick Click handler to be called on uploads emanating from folders
 * @return {JSX}
 */
export default function FileLI({
  bucket,
  type,
  name,
  lastModified = null,
  size = null,
  filePath = null,
  folderClick,
  locale,
  uploadClick = null,
}) {
  // For Both
  const [toggleTooltip, setToggleTooltip] = useState(false);
  // For Folders
  const [toggleDeleteFolderModal, setToggleDeleteFolderModal] = useState(false);

  // For Files
  const [toggleFileModal, setToggleFileModal] = useState(false);

  function onKeyDownFolder(e) {
    e.preventDefault();
    if (e.key === 'Enter') folderClick(filePath);
  }

  // *********** Folder *********** //
  function handleToggleTooltip() {
    setToggleTooltip((s) => !s);
  }

  function handleCloseTooltip() {
    setToggleTooltip(false);
  }

  function handleDeleteModal() {
    setToggleDeleteFolderModal((s) => !s);
  }

  function formatLastModified(timeString) {
    const timeArray = timeString.split(/[T.]/);
    return `${timeArray[0]} ${timeArray[1]}`;
  }

  if (type === 'folder') {
    return (
      <motion.li
        layout
        className="file-folder-li"
      >
        <div className="file-row-wrapper">
          <div className="file-row">
            <div
              role="button"
              tabIndex={0}
              onClick={() => folderClick(filePath)}
              onKeyDown={(e) => onKeyDownFolder(e)}
              className="file-row-left"
              data-path={filePath}
              data-type="folder"
            >
              <div className="icon-wrapper"><Folder /></div>
              <p className="file-table-data">{name}</p>
            </div>
            <IoEllipsisVerticalCircle
              className="file-row-right"
              onClick={handleToggleTooltip}
            />
          </div>
          <FolderDropdown
            isOpen={toggleTooltip}
            toggleDeleteFolderModal={handleDeleteModal}
            closeDropdown={handleToggleTooltip}
            buttonClick={handleCloseTooltip}
            uploadClick={() => uploadClick(filePath)}
          />
          { toggleDeleteFolderModal
    && (
    <ModalComponentWrapper close={handleDeleteModal}>
      <FolderDeleteModal
        locale={locale}
        bucket={bucket}
        pathToDelete={filePath}
        folderName={name}
      />
    </ModalComponentWrapper>
    ) }
        </div>
      </motion.li>
    );
  }

  // ************ File ************ //
  function handleToggleFileModal() {
    setToggleFileModal((s) => !s);
  }
  const downloadInfo = { locale, bucket, filePath };
  const fileInfo = { name, lastModified, size };
  return (
    <motion.li
      layout
      className="file-folder-li"
    >
      <div className="file-row-wrapper">
        <div
          className="file-row"
        >
          <div
            className="file-row-left"
            role="button"
            tabIndex={0}
            data-path={filePath}
            data-type="file"
          >
            <div className="icon-wrapper"><FileIcon name={name} /></div>
            <p className="file-table-data name">{name}</p>
            <p className="file-table-data last-modified">{() => formatLastModified(lastModified)}</p>
            <p className="file-table-data size">{() => fileSizeTruncate(size)}</p>
          </div>
          <IoInformationCircleSharp
            className="file-row-right"
            onClick={() => handleToggleFileModal()}
          />
        </div>
        {toggleFileModal
        && (
        <ModalComponentWrapper close={handleToggleFileModal}>
          <FileModal
            fileInfo={fileInfo}
            downloadInfo={downloadInfo}
          />
        </ModalComponentWrapper>
        )}
      </div>
    </motion.li>
  );
}

FileLI.defaultProps = {
  bucket: '',
  type: '',
  name: '',
  lastModified: null,
  size: null,
  filePath: null,
  folderClick: null,
  locale: '',
  uploadClick: () => null,
};

FileLI.propTypes = {
  bucket: propTypes.string,
  type: propTypes.string,
  name: propTypes.string,
  lastModified: propTypes.string,
  size: propTypes.number,
  filePath: propTypes.string,
  folderClick: propTypes.func,
  locale: propTypes.string,
  uploadClick: propTypes.func,
};
