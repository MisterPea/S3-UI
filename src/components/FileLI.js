import * as React from 'react';
import propTypes from 'prop-types';
import { Folder } from './graphic_elements/Icons';
import FileIcon from './helpers/FileIcon';

/**
 * Component to render a folder or a file
 * @prop {string} type The type to be rendered `file` or `folder`
 * @prop {string} name Name associated with the file or folder
 * @prop {string} lastModified Date object was last modified (Only for files)
 * @prop {number} size Size of the file (Only for files)
 * @prop {string} filePath File path leading to children (Only for folders)
 * @return {JSX}
 */
export default function FileLI({
  type, name, lastModified = null, size = null, filePath = null, callback,
}) {
  function onKeyDownFolder(e) {
    e.preventDefault();
    if (e.key === 'Enter') callback(filePath);
  }

  if (type === 'folder') {
    return (
      <div
        className="file-row"
        role="button"
        tabIndex={0}
        onClick={() => callback(filePath)}
        onKeyDown={(e) => onKeyDownFolder(e)}
      >
        <div className="icon-wrapper"><Folder /></div>
        <p className="file-table-data">{name}</p>
      </div>
    );
  }
  return (
    <div
      className="file-row"
    >
      <div className="icon-wrapper"><FileIcon name={name} /></div>
      <p className="file-table-data name">{name}</p>
      <p className="file-table-data last-modified">{lastModified}</p>
      <p className="file-table-data size">{size}</p>
      <div className="file-table-data options">Options</div>
    </div>
  );
}

FileLI.defaultProps = {
  type: '',
  name: '',
  lastModified: null,
  size: null,
  filePath: null,
  callback: null,
};

FileLI.propTypes = {
  type: propTypes.string,
  name: propTypes.string,
  lastModified: propTypes.string,
  size: propTypes.number,
  filePath: propTypes.string,
  callback: propTypes.func,
};
