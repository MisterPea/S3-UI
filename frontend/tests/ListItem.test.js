import * as React from 'react';
import ReactDOM from 'react-dom';
import toJson from 'enzyme-to-json';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FileLI from '../src/components/FileLI';

configure({ adapter: new Adapter() });

describe('FileLI component', () => {
  it('Renders without crashing', () => {
    const div = document.createElement('DIV');
    ReactDOM.render(<FileLI type={'folder'} name={'empty folder'}/>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Renders the folder info passed to it', () => {
    const wrapper = shallow(<FileLI type={'folder'} name={'empty folder'}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
