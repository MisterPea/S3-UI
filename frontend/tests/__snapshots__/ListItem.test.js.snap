// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileLI component Renders the folder info passed to it 1`] = `
<div
  className="file-row"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="icon-wrapper"
  >
    <Folder />
  </div>
  <p
    className="file-table-data"
  >
    empty folder
  </p>
</div>
`;
